### ISSUE TITLE
trimesh.Trimesh.section weird precision problem

### ISSUE THREAD

**Issue by JohnYeung-dojjy** [mikedh/trimesh#2121]
Hi, I was trying to use mesh.section to retrive some closed region on a mesh, but in 1 out of 1000 mesh, it seems there are some precision errors, although this should be handled in the unique_row functions.
```python
slice3D = mesh.section(
    plane_normal=plane.normal,
    plane_origin=plane.normal * plane.constant
)
slice2D, _ = slice3D.to_planar()
print(slice2D.entities)
slice2D.plot_entities()
```

![image](https://github.com/mikedh/trimesh/assets/62557955/7ec8203d-ead3-4c58-90be-99af2d6fdee1)
When I used a custom function (preform rounding to lines before passing into `load_paths()`), the issue was gone
```python
def section_round(
        mesh: trimesh.Trimesh, plane_normal: list[int], plane_origin: list[int], **kwargs
    ) -> Path3D:
        """
        Returns a 3D cross section of the current mesh and a plane
        defined by origin and normal.

        Parameters
        ------------
        plane_normal: (3) vector for plane normal
          Normal vector of section plane
        plane_origin : (3, ) float
          Point on the cross section plane

        Returns
        ---------
        intersections: Path3D or None
          Curve of intersection
        """
        # turn line segments into Path2D/Path3D objects
        from trimesh.exchange.load import load_path

        # return a single cross section in 3D
        lines, face_index = trimesh.intersections.mesh_plane(
            mesh=mesh,
            plane_normal=plane_normal,
            plane_origin=plane_origin,
            return_faces=True,
            **kwargs,
        )

        # if the section didn't hit the mesh return None
        if len(lines) == 0:
            return None

        # otherwise load the line segments into a Path3D object
        path = load_path(lines.round(13)) # only this line was changed

        # add the face index info into metadata
        path.metadata["face_index"] = face_index

        return path
```
```python
slice3D = section_round(
    mesh,
    plane_normal=plane.normal,
    plane_origin=plane.normal * plane.constant
)
slice2D, _ = slice3D.to_planar()
print(slice2D.entities)
slice2D.plot_entities()
```

![image](https://github.com/mikedh/trimesh/assets/62557955/c2fa3214-9f44-4257-b856-afab8fa25aea)

I suspect it is related to `hashable_rows()` called in `unique_rows()`, what do you think?
![image](https://github.com/mikedh/trimesh/assets/62557955/28482159-fbf4-43fd-86ae-e4a9a9971e9b)

**Comment by JohnYeung-dojjy** [comment]
and if I call `slice2D.discrete`, it gives an empty array, which is not the case when I swap to my custom `section_round()`

**Comment by mikedh** [comment]
Does it also fix it if you try different values for the merge tolerance?

```
In [1]: import trimesh

In [3]: trimesh.constants.tol_path.merge
Out[3]: 1e-05

In [4]: trimesh.constants.tol.merge
Out[4]: 1e-08

In [5]: trimesh.constants.tol.merge = 1e-13

In [6]: trimesh.constants.tol_path.merge = 1e-13


```

**Comment by JohnYeung-dojjy** [comment]
Thank you for the prompt reply, I will have a try today.

**Comment by JohnYeung-dojjy** [comment]
Although I am unable to find which mesh had the said issue (because I forgot to log it's id :man_facepalming:)
But it does solve other meshes with similar issues.
```python
# the vertices precision is about 8~10 d.p.
# tol too large, the algorithm might confuse between 2 distinct points
# tol too small, the algorithm will be prompt to floating point precision error
trimesh.constants.tol_path.merge = 1e-13
trimesh.constants.tol.merge      = 1e-13
```

I believe setting tol to 1e-13 can handle most of the edge cases, thank you!

**Comment by JohnYeung-dojjy** [comment]
Update, I have found the original mesh that has the precision issue.
It can be solved by setting tol to 1e-10
1e-13 will cause a complete path to split into several paths.
It seems to me that to approach to find unique rows by hashing int(scaled-float) made the algorithm sensitive to precision issues

**Comment by JohnYeung-dojjy** [comment]
Final update:

# TLDR
set a smaller merge tol, then apply Path.fill_gaps()

# Previous discussion does not work very well
Choosing the situatable tol values is not enough, because it is prompt to precision issues if you set it too high or too small

# The problem that I face
In my use case, I want to obtain a continous linepoints that defines the mesh-plane intersection. unfortunately for a small variance in the merge tol I set.

## Setting the merge tol too large
If I set it soo large, the path.discrete can disconnect / path is connect to an opposite end for some reason

### Example mesh that the path is connected to opposite end (not to scale)
![image](https://github.com/mikedh/trimesh/assets/62557955/64f79136-1beb-4e78-bcae-f061b69f6b1e)

## Setting the merge tol too small
If I set it soo small, the connected paths are disconnected even though they should.

### Example mesh the the paths are disconnected
![image](https://github.com/mikedh/trimesh/assets/62557955/3e728f77-fa6a-4467-9879-03af647f1362)

These problems happens to a small portion of the meshes I worked with.

# Working solution
Likely I have found that mixing the use of setting merge tol along with the use of Path.fill_gaps() can solve my problem.
I still don't know why setting a larger merge tol can lead to such error. But if I set a smaller merge tol, Path.fill_gaps() can reconnect the path endpoints that should have been connected (because they are in fact the same vertex, with <1e-8 floating point precision difference).
